### PR TITLE
Recursive load_state_dict()

### PR DESCRIPTION
**Pull request by mys007**
I propose to make tensor assignments in `load_state_dict()` recursive over the tree of modules. The advantage of this somewhat more complicated code is that one is able to handle versioning of `state_dict` easily in an encapsulated way inside each specific `Module`. For example:

```
def load_state_dict(self, state_dict, prefix, state_tree):
    if 'member1' in state_tree: # renaming
        state_tree['renamed_member'] = state_tree['member1']
        del state_tree['member1']
    if `member2` in state_tree: # removing
        del state_tree['member2']
    if `member3` not in state_tree: #adding
        state_tree['member3'] = torch.zeros(5,5)
    super(Foo, self).load_state_dict(state_dict, prefix, state_tree)            
```

### CODE REVIEW

**Comment by apaszke** [comment]
Can you please explain what's the difference between `state_dict` and `state_tree`? I don't understand the code example in the PR

**Comment by mys007** [comment]
Sorry for the convolved code. 
`state_dict` has exactly the same meaning as in the original code (keys have hierarchy encoded in dot-notation, e.g. `{'1.1.weight': A, '1.1.bias': B, '2.weight': C}`). 
`state_tree` is a tree built from `state_dict` using `_fill_tree()`, where each tree node is a dictionary (inner node) or a tensor (leaf node), e.g.
```
{
  '1': {
         '1': {'weight': A, 'bias': B}
       },
  '2': {'weight': C}
}
```
The proposed implementation in `nn.Module` calls child's `load_state_dict()` with the the respective branch. In the motivational code snippet above, a module (e.g. '1.1') uses `state_tree` (e.g. `{'weight': A, 'bias': B}`) to modify the default loading behavior without the need to know about its global position in the model.

**Comment by apaszke** [comment]
I think this would be a really confusing API. Do I need to pass both the state dict and state tree? Why aren't they exclusive arguments? When do I pick one or another?

I don't think it's worth adding this in here, just so that we can keep things simple. `load_state_dict` isn't meant to be a general super powerful network transforming function, but to restore the state to what `state_dict` returned.

Anyway, thanks a lot for the PR, but please discuss such changes with us in an issue before you send it!

**Comment by mys007** [comment]
The API would not change w.r.t. to the user at all, she would still just make `model.load_state_dict(checkpoint['state_dict'])` kind of a call as before. It's only when overloading `load_state_dict()` of a specific module (which is not done currently because it IMHO doesn't help to solve any problem) where the change would be seen. The motivation is to solve cases where the module definition has changed and the mapping between current and saved states doesn't match anymore. For instance, the problem addressed in https://github.com/pytorch/pytorch/pull/2815 could be solved here easily and safely. But anyway, thanks for your time considering the PR.

**Comment by soumith** [comment]
@mys007 martin, to give you additional context. We designed state_dict / load_state_dict to be explicitly stupid and easily manipulatable, so we wouldn't really like additional complexity added to it.
